### PR TITLE
An end-to-end training example

### PR DESCRIPTION
**Pull request by ZongyuLi-umich**
Will add an GPU version example under `CuSPECTrecon.jl`.

### CODE REVIEW

File: docs/lit/examples/6-dl.jl [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+# We set β = 1 and train 2 outer iterations.
+β = 1
+Asum = A' * ones(T, nx, nz, nview)
+scatters = scatter_mean * ones(T, nx, nz, nview)
+function loss(xrecon, xtrue)
+    xiter1 = bregem(projectb, backprojectb, ynoisy, scatters,
+				    Asum, xrecon, cnn, β; niter = 1)
+    xiter2 = bregem(projectb, backprojectb, ynoisy, scatters,
+				    Asum, xiter1, cnn, β; niter = 1)
+    return sum((xiter2 .- xtrue).^2)
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:207]:
```suggestion
    return sum(abs2, xiter2 - xtrue)
```

**Comment by JeffFessler** [comment]
This is really a great addition!  Just a couple minor things to address first.

**Comment by ZongyuLi-umich** [comment]
Thanks for your comments! See the updated code.

File: docs/lit/examples/6-dl.jl [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+# We set β = 1 and train 2 outer iterations.
+β = 1
+Asum = A' * ones(T, nx, nz, nview)
+scatters = scatter_mean * ones(T, nx, nz, nview)
+function loss(xrecon, xtrue)
+    xiter1 = bregem(projectb, backprojectb, ynoisy, scatters,
+				    Asum, xrecon, cnn, β; niter = 1)
+    xiter2 = bregem(projectb, backprojectb, ynoisy, scatters,
+				    Asum, xiter1, cnn, β; niter = 1)
+    return sum(abs2, xiter2 - xtrue)
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:207]:
```suggestion
    return sqrt(sum(abs2, xiter2 - xtrue) / sum(abs2, xtrue)) # NRMSE
```

File: docs/lit/examples/6-dl.jl [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+# This page describes how to end-to-end train unrolled deep learning algorithms
+# using the Julia package
+# [`SPECTrecon`](https://github.com/JeffFessler/SPECTrecon.jl).
+
+
+# ### Setup
+
+# Packages needed here.
+
+using LinearAlgebra
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:14]:
```suggestion
using LinearAlgebra: norm, mul!
```

File: docs/lit/examples/6-dl.jl [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+using LinearMapsAA: LinearMapAA
+using LinearAlgebra: mul!
+forw! = (y,x) -> project!(y, x, plan)
+back! = (x,y) -> backproject!(x, y, plan)
+idim = (nx,ny,nz)
+odim = (nx,nz,nview)
+A = LinearMapAA(forw!, back!, (prod(odim),prod(idim)); T, odim, idim)
+
+# Generate noisy data
+using Distributions: Poisson
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:89]:
```suggestion
```

File: docs/lit/examples/6-dl.jl [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+"""
+function bregem(projectb::Function,
+                backprojectb::Function,
+                y::AbstractArray,
+                r::AbstractArray,
+                Asum::AbstractArray,
+                x::AbstractArray,
+                cnn::Chain,
+                β::Real;
+                niter::Int = 1)
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:183]:
```suggestion
function bregem(
    projectb::Function,
    backprojectb::Function,
    y::AbstractArray,
    r::AbstractArray,
    Asum::AbstractArray,
    x::AbstractArray,
    cnn::Chain,
    β::Real;
    niter::Int = 1,
)
```

File: docs/lit/examples/6-dl.jl [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+using LinearAlgebra: norm, mul!
+using SPECTrecon
+using MIRTjim: jim, prompt
+using Plots: scatter, plot!, default; default(markerstrokecolor=:auto)
+using Zygote
+using ZygoteRules
+using Flux
+using LinearMapsAA: LinearMapAA
+using Distributions: Poisson
+
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:23]:
```suggestion
using BSON: @load
```

Comment (JeffFessler) [reply docs/lit/examples/6-dl.jl:23]: we also need BSON in the Project.toml!

File: docs/lit/examples/6-dl.jl [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+##     opt = ADAMW(0.002)
+##     Flux.Optimise.update!(opt, ps, gs)
+## end
+
+# Uncomment to save your trained model
+## using BSON: @save
+## @save "../data/trained-cnn-example-6-dl.bson" cnn
+
+# load the pretrained model
+using BSON: @load
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:232]:
```suggestion
```

File: docs/lit/examples/6-dl.jl [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+##	   ps = Flux.params(cnn)
+##     gs = gradient(ps) do
+##         loss(xhat1, xtrue) # we start with the 30 iteration EM reconstruction
+##     end
+##     opt = ADAMW(0.002)
+##     Flux.Optimise.update!(opt, ps, gs)
+## end
+
+# Uncomment to save your trained model
+## using BSON: @save
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:228]:
```suggestion
```

File: docs/Project.toml [hunk 9]
```diff
@@ -1,14 +1,18 @@
...
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 LinearMapsAA = "599c1a8e-b958-11e9-0d14-b1e6b2ecea07"
 Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
 MIRTjim = "170b2178-6dee-4cb0-8729-b3e8b57834cc"
 Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
 SPECTrecon = "ab1be465-a7f0-4423-9048-0ee774b70ed9"
 Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
 UnitfulRecipes = "42071c24-d89e-48dd-8a24-8a12d9b8861f"
+Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
+ZygoteRules = "700de1a5-db45-46bc-99cf-38207098b444"
```
Comment (JeffFessler) [inline docs/Project.toml:16]:
```suggestion
BSON = "fbb218c0-5317-5bc6-957e-2ee96dd4b1f0"
```

**Pull request merged**